Move the shape env symint cache to a symbol cache, better routing for subclass fakification

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115227
* #114965
* #115212
* #114311

Context:

Joel sees that unless he manually writes to the fake tensor memo, fakification seems to produce spurious symbols! Voz (me) objects, saying that not only is directly writing to memo a bad pattern, recursively invoking fakification on tensor subclass elements in dynamo should suffice! Joel says that while he morally agrees, he has a test proving otherwise, a most perplexing situation.

Digging in, I figured out that while *we were* making fake tensors correctly, with properly cached symbols and the like, we were *also* incorrectly creating spurious symbols, leading the test to fail.

Before this PR, we would only cache source->symint. This was generally fine, but meant that you would create a symbol, then potentially throw it out due to symint cache. For example, the cache hit flow was:

make a symbol (ex: s2) -> use it to make a symint -> hit the cache (my_source-s1)

Now, in this example,  you have a symbol in your val_to_var/var_to_val (s2) that is unused. This is sound, but wasteful, and furthermore, misleading.

This was causing a test added in a PR in this stack to fail, specifically, because the test was using

```
curr_var_to_val = {
    str(k): v for k, v in context.fake_mode.shape_env.var_to_val.items()
}
````

To validate that no new symbols were being created (that is, that recursively creating fake tensors for subclasses was working).

The test is correct, but the implementation of caching would make (by this method of observation) cache hits look like cache misses.

So, the fix here is to move the cache up to be a general symbol cache, rather than only a cache for symints.

The initial implementation did that! But then, it ran into some interesting errors when it came to replay. When replaying symbol creation, behaviors would diverge in the new shape env! How could that be? The answer is because creating a new shape_env resulted in us replaying symbol creation... but with a cache from a different shape env! This was short circuiting symbol creation - and so, adding an extra layer to the cache for id(shape_env) fixes the problem. 


cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

voznesenskym committed:
Move the shape env symint cache to a symbol cache, better routing for subclass fakification

[ghstack-poisoned]

voznesenskym committed:
Update on "Move the shape env symint cache to a symbol cache, better routing for subclass fakification"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "Move the shape env symint cache to a symbol cache, better routing for subclass fakification"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "Move the shape env symint cache to a symbol cache, better routing for subclass fakification"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1797,7 +1799,12 @@ def wrap_to_fake_tensor_and_record(e, tx, *, source: Optional[Source], is_tensor
         )
 
-        symbolic_context = _automatic_dynamic(e, tx, source, static_shapes)
+        if not parent_context:
```
self todo: comment on identical recursion pattern vs parent

jbschlosser commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1820,18 +1827,15 @@ def wrap_to_fake_tensor_and_record(e, tx, *, source: Optional[Source], is_tensor
-                # no need to fake-ify the inner tensors again later on
-                tx.fake_mode.fake_tensor_converter.set_tensor_memo(inner, fake_inner)
-
         for fake, real, source, symbolic_context in tracking_info:
```
nit: now that `tracking_info` is static above, we can simplify by removing it -> the iteration here becomes unnecessary

jbschlosser commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -858,13 +858,13 @@ class StatefulSymbolicContext(StatelessSymbolicContext):
     """
     tensor_source: Source = None
-    source_to_symint_node_cache : Dict["TensorPropertySource", SymInt] = None
+    shape_env_to_source_to_symbol_cache : Dict[int, Dict["TensorPropertySource", "sympy.Expr"]] = None
```
nit: maybe add a comment mentioning that the int key is id(shape_env)

jbschlosser commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2375,7 +2370,7 @@ def create_unspecified_symbol(
+        return self.create_symbol(val, source, dynamic_dim, constraint_dim, positive=None, do_not_specialize_zero_one=True, symbolic_context=None)
 
     @record_shapeenv_event()
     def create_symbol(
```
so now we're caching at the `create_symbol()` level instead of at the `create_symintnode()` level, correct?

does `create_symintnode()` still need to take in a `symbolic_context`? doesn't look like it's used

voznesenskym replied:
rm

jbschlosser reviewed (approved):
cool thanks!

voznesenskym committed:
Update on "Move the shape env symint cache to a symbol cache, better routing for subclass fakification"


Context:

Joel sees that unless he manually writes to the fake tensor memo, fakification seems to produce spurious symbols! Voz (me) objects, saying that not only is directly writing to memo a bad pattern, recursively invoking fakification on tensor subclass elements in dynamo should suffice! Joel says that while he morally agrees, he has a test proving otherwise, a most perplexing situation.

Digging in, I figured out that while *we were* making fake tensors correctly, with properly cached symbols and the like, we were *also* incorrectly creating spurious symbols, leading the test to fail.

Before this PR, we would only cache source->symint. This was generally fine, but meant that you would create a symbol, then potentially throw it out due to symint cache. For example, the cache hit flow was:

make a symbol (ex: s2) -> use it to make a symint -> hit the cache (my_source-s1)

Now, in this example,  you have a symbol in your val_to_var/var_to_val (s2) that is unused. This is sound, but wasteful, and furthermore, misleading.

This was causing a test added in a PR in this stack to fail, specifically, because the test was using

```
curr_var_to_val = {
    str(k): v for k, v in context.fake_mode.shape_env.var_to_val.items()
}
````

To validate that no new symbols were being created (that is, that recursively creating fake tensors for subclasses was working).

The test is correct, but the implementation of caching would make (by this method of observation) cache hits look like cache misses.

So, the fix here is to move the cache up to be a general symbol cache, rather than only a cache for symints.

The initial implementation did that! But then, it ran into some interesting errors when it came to replay. When replaying symbol creation, behaviors would diverge in the new shape env! How could that be? The answer is because creating a new shape_env resulted in us replaying symbol creation... but with a cache from a different shape env! This was short circuiting symbol creation - and so, adding an extra layer to the cache for id(shape_env) fixes the problem. 


cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -858,13 +858,13 @@ class StatefulSymbolicContext(StatelessSymbolicContext):
     """
     tensor_source: Source = None
-    source_to_symint_node_cache : Dict["TensorPropertySource", SymInt] = None
+    shape_env_to_source_to_symbol_cache : Dict[int, Dict["TensorPropertySource", "sympy.Expr"]] = None
```
Question around invariants: when do we have more than one ShapeEnv floating around? Is there a test I can look at that exercises the fact that assuming a single ShapeEnv during this caching is wrong?

It seems a bit weird to me that we have to include a ShapeEnv id in this cache lookup, since:

(1) we should be able to get away with a single ShapeEnv throughout the lifecycle of tracing + compiling a single dynamo graph

(2) I would have thought that this cache (and the tracing context that holds it) would get reset after every dynamo graph break, when start tracing out a fresh graph

voznesenskym replied:
Check the PR desc :) 

>  Is there a test I can look at that exercises the fact that assuming a single ShapeEnv during this caching is wrong?

`python test/dynamo/test_dynamic_shapes.py -k test_autocast_arguments_binding_dynamic_shapes`

Should do it!

```
So, the fix here is to move the cache up to be a general symbol cache, rather than only a cache for symints.

The initial implementation did that! But then, it ran into some interesting errors when it came to replay. When replaying symbol creation, behaviors would diverge in the new shape env! How could that be? The answer is because creating a new shape_env resulted in us replaying symbol creation... but with a cache from a different shape env! This was short circuiting symbol creation - and so, adding an extra layer to the cache for id(shape_env) fixes the problem.
```

(1) Correct, but we replay for correctness
(2) It does :)

bdhirsh replied:
When (/ why) do we replay symbol creation? I'll read through the code some more.

> python test/dynamo/test_dynamic_shapes.py -k test_autocast_arguments_binding_dynamic_shapes

So, I tried running this test locally, but "deleting" the per-shape-env cache, just to understand how it would fail. And it still passes for me? I deleted the per-shape-env bits by just changing the `id(self)` key (the id of the ShapeEnv) to 1 everywhere (lol) https://www.internalfb.com/phabricator/paste/view/P900020748

voznesenskym committed:
Update on "Move the shape env symint cache to a symbol cache, better routing for subclass fakification"


Context:

Joel sees that unless he manually writes to the fake tensor memo, fakification seems to produce spurious symbols! Voz (me) objects, saying that not only is directly writing to memo a bad pattern, recursively invoking fakification on tensor subclass elements in dynamo should suffice! Joel says that while he morally agrees, he has a test proving otherwise, a most perplexing situation.

Digging in, I figured out that while *we were* making fake tensors correctly, with properly cached symbols and the like, we were *also* incorrectly creating spurious symbols, leading the test to fail.

Before this PR, we would only cache source->symint. This was generally fine, but meant that you would create a symbol, then potentially throw it out due to symint cache. For example, the cache hit flow was:

make a symbol (ex: s2) -> use it to make a symint -> hit the cache (my_source-s1)

Now, in this example,  you have a symbol in your val_to_var/var_to_val (s2) that is unused. This is sound, but wasteful, and furthermore, misleading.

This was causing a test added in a PR in this stack to fail, specifically, because the test was using

```
curr_var_to_val = {
    str(k): v for k, v in context.fake_mode.shape_env.var_to_val.items()
}
````

To validate that no new symbols were being created (that is, that recursively creating fake tensors for subclasses was working).

The test is correct, but the implementation of caching would make (by this method of observation) cache hits look like cache misses.

So, the fix here is to move the cache up to be a general symbol cache, rather than only a cache for symints.

The initial implementation did that! But then, it ran into some interesting errors when it came to replay. When replaying symbol creation, behaviors would diverge in the new shape env! How could that be? The answer is because creating a new shape_env resulted in us replaying symbol creation... but with a cache from a different shape env! This was short circuiting symbol creation - and so, adding an extra layer to the cache for id(shape_env) fixes the problem. 


cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge -f "Ghstack is wrong or something"

voznesenskym commented:
> ## Merge failed
> **Reason**: PR 115227 is out of sync with the corresponding revision [169c736](https://github.com/pytorch/pytorch/commit/169c736d45524c2e80d38a57081de3dcfcaced9e) on branch origin/gh/voznesenskym/288/orig that would be merged into main. This usually happens because there is a non ghstack change in the PR. Please sync them and try again (ex. make the changes on origin/gh/voznesenskym/288/orig and run ghstack).
> 
> Details for Dev Infra team

I wonder if this is because a push failed due to a github 500 earlier?

voznesenskym commented:
> ## Merge failed
> **Reason**: PR 115227 is out of sync with the corresponding revision [169c736](https://github.com/pytorch/pytorch/commit/169c736d45524c2e80d38a57081de3dcfcaced9e) on branch origin/gh/voznesenskym/288/orig that would be merged into main. This usually happens because there is a non ghstack change in the PR. Please sync them and try again (ex. make the changes on origin/gh/voznesenskym/288/orig and run ghstack).
> 
> Details for Dev Infra team

https://github.com/pytorch/pytorch/pull/115396
